doc/docs/recipe-versioning.md

```diff
@@ -0,0 +1,83 @@
+---
+title: Versioning Data
+slug: /recipes/versioning-data
+---
+
+# Versioning in Replicache
+
+There are a few versions you as a developer needs to be aware of when working
+with Replicache.
+
+First there is the _Replicache format version_ which describes the data that
+Replicache stores persistently, in Indexed DB at the moment, but it should be
+treated as an implementation detail. This version is not exposed by the
+Replicache API, but when Replicache changes its on disk format this version is
+also changed. Whenever this changes it is a breaking change and requires a new
+major release of Replicache.
+
+Then there is the
+[schemaVersion](../api/interfaces/ReplicacheOptions#schemaversion). This is
+optional but it is recommended that you always provide this. It is used to
+describe the format of your [mutators](../api/interfaces/ReplicacheOptions#mutators)
+as well as your [client view](/guide/client-view).
+
+Finally there is the [pullVersion](/server-pull#pullversion) and
+[pushVersion](/server-push#pushversion). These are used to describe the response
+format of the pull and push requests.
+
+### schemaVersion
+
+The `schemaVersion` version is something you control and it is used to describe
+that shape of the data that the client expects as well as the signatures of the
+mutators your client and server expects.
+
+If you want to change the arguments passed to a mutator in a non backwards way
+you have two options:
+
+1. Keep the existing mutator as is and add a new mutator with a new name. On the
+   server you need to keep the old mutator with the old name to support old
+   clients (until there are no more old clients)
+2. Change the mutator and change the `schemaVersion`. In your push handler on
+   the server you now need to check the `schemaVersion` to know what mutator
+   arguments the client sent.
+
+If the data sent down in a pull response changes in a non backwards way you need
+to change the `schemaVersion` passed into the Replicache constructor. The server
+then needs to check what version the client is expecting and reply with data the
+client understands. If the server no longer supports the client's version it can
+reply with an HTTP error code. If the server replies with an error the client
+will not see any new changes and the client needs to be updated. At the moment
+we do not handle this case for you but in the future we are [planning to
+add](https://github.com/rocicorp/replicache/issues/335) a way for the server to
+tell the client that the data changed in a way that it has to start over.
+
+### pullVersion
+
+`pullVersion` is used to describe the format of the response of the pull
+request. It is currently at `0`. You should check that the response of the pull
```

Comment:
WDYT about removing this feature @grgbkr @phritz. If we ever change the format of the pull request in a backward incompatible way, we will change the major version of Replicache itself (so devs won't get the update accidentally) and we can add this feature at that time and explain how to use it in release notes. Now I feel it creates extra complexity to understand for no benefit. Same with `pushVersion`

Reply:
I like this idea.  Simplify!

Reply:
> for no benefit

The benefit is that _if_ we change the pull request in a backwards-incompatible way _then_ a server that does not yet speak that new version won't try and fail or do the wrong thing when it sees the new kind of pull request that it doesn't understand. If we remove pullVersion then existing servers yet to be updated to the new thing have no way to know that they are receiving a payload they don't expect and can't handle. We could let them fail unpredictably or do the wrong thing but seems nicer to be able to error out with a clean, clear message. We could also _require_ that customers always upgrade their servers before moving to a new client, or do it in lockstep, but that seems brittle.

I'm open to removing it if people think the benefit is not worth the cost.

Reply:
(To be clear, I don't care about this particular mechanism, I just care that there is a mechanism to ensure the client doesn't send the server something that is easy for it to misinterpret. If we stick version info in the path or something that works too (eg `/api/v2/push` or similar)).